src/osd/PrimaryLogPG.cc

```diff
@@ -6606,33 +6645,55 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
 	  result = -EINVAL;
 	  break;
 	}
-        oi.set_flag(object_info_t::FLAG_MANIFEST);
-	oi.manifest.redirect_target = target;
-	oi.manifest.type = object_manifest_t::TYPE_REDIRECT;
-	t->truncate(soid, 0);
-	if (oi.is_omap() && pool.info.supports_omap()) {
-	  t->omap_clear(soid);
-	  obs.oi.clear_omap_digest();
-	  obs.oi.clear_flag(object_info_t::FLAG_OMAP);
-	}
-        ctx->delta_stats.num_bytes -= oi.size;
-	oi.size = 0;
-	oi.new_object();
-	oi.user_version = target_version;
-	ctx->user_at_version = target_version;
-	/* rm_attrs */
-	map<string,bufferlist> rmattrs;
-	result = getattrs_maybe_cache(ctx->obc,
-		    &rmattrs);
-	if (result < 0) {
-	  return result;
-	}
-	map<string, bufferlist>::iterator iter;
-	for (iter = rmattrs.begin(); iter != rmattrs.end(); ++iter) {
-	  const string& name = iter->first;
-	  t->rmattr(soid, name);
+	bool need_reference = osd_op.op.flags & CEPH_OSD_OP_FLAG_WITH_REFERENCE ||
+			      oi.flags & object_info_t::FLAG_REDIRECT_HAS_REFERENCE;
```

Comment:
both of them

Reply:
I think this needs_reference check breaks if the object is already a redirect to object A (and has a ref) and we are setting it to point to object B.  Perhaps set_redirect should fail if the object is already a redirect to a different object?

Reply:
Ideally the librados user could do a compound operation that unsets the old ref and then sets the new one.. and we'd still do the ref put at the end?

Reply:
Also, I don't see why the object already being a redirect that owns a reference implies that the *new* redirect we are setting should also take a reference...

Reply:
It (check breaks if the object is already a redirect) would be simpler than before at this point. Fixed.
Also, I will make a new PR that implement a compound operation (unset + set) as your comment.